File: Data-Structure-I/20-Valid-Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        if(s.length()%2!=0) return false;
        stack<char> st;
        for(char ch : s){
            if(ch=='}' || ch==']' || ch==')'){
                if(st.empty()) return false;
                else if(ch==']' && st.top()!='[') return false;
                else if(ch=='}' && st.top()!='{') return false;
                else if(ch==')' && st.top()!='(') return false;
                else st.pop();
            }
            else
                st.push(ch);
        }
        return (st.empty()) ? true : false;
    }
};

File: Data-Structure-I/20-Valid-Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        if len(s)%2 == 1:
            return False
        for ch in s:
            if ch in "]})":
                if not st:
                    return False
                elif ch==')' and st[-1]!='(':
                    return False
                elif ch==']' and st[-1]!='[':
                    return False
                elif ch=='}' and st[-1]!='{':
                    return False
                else:
                    st.pop()
            else:
                st.append(ch)
        return True if not st else False
